griptape/loaders/base_loader.py

```diff
@@ -18,8 +20,23 @@ class BaseLoader(ABC):
     def load(self, source: Any, *args, **kwargs) -> BaseArtifact | Sequence[BaseArtifact]:
         ...
 
-    @abstractmethod
     def load_collection(
-        self, sources: list[Any], *args, **kwargs
+        self, sources: Sequence[Any], *args, **kwargs
     ) -> Mapping[str, BaseArtifact | Sequence[BaseArtifact | Sequence[BaseArtifact]]]:
-        ...
+        # Create a dictionary before actually submitting the jobs to the executor
+        # to avoid duplicate work.
+        sources_by_key = {self.to_key(source): source for source in sources}
+        return execute_futures_dict(
+            {
+                key: self.futures_executor.submit(self.load, source, *args, **kwargs)
+                for key, source in sources_by_key.items()
+            }
+        )
+
+    def to_key(self, source: Any, *args, **kwargs) -> str:
```

Comment:
Should we rename to `to_hash`?

Reply:
We could, but it doesn't necessarily need to be a hash. It just need to be the same value as the key in the dict returned by `load_collection`. Let me know if you feel strongly and I'll change it

Reply:
Nah its fine